src/ImmersedBoundaries/active_cells_map.jl

```diff
@@ -1,50 +1,121 @@
-@inline use_only_active_cells(grid::ActiveCellsIBG) = true
+struct InteriorMap end
+struct SurfaceMap end
 
-@inline active_cells_work_layout(size, grid::ActiveCellsIBG) = min(length(grid.active_cells_map), 256), length(grid.active_cells_map)
-@inline active_linear_index_to_ntuple(idx, grid::ActiveCellsIBG) = Base.map(Int, grid.active_cells_map[idx])
+@inline use_only_active_interior_cells(grid::ActiveCellsIBG) = InteriorMap()
 
-function ImmersedBoundaryGrid{TX, TY, TZ}(grid, ib; active_cells_map = false) where {TX, TY, TZ} 
+@inline use_only_active_surface_cells(grid::AbstractGrid)   = nothing
+@inline use_only_active_surface_cells(grid::ActiveCellsIBG) = SurfaceMap()
 
+@inline active_cells_work_layout(group, size, ::InteriorMap, grid::ActiveCellsIBG) = min(length(grid.active_cells_interior), 256), length(grid.active_cells_interior)
+@inline active_cells_work_layout(group, size, ::SurfaceMap,  grid::ActiveCellsIBG) = min(length(grid.active_cells_surface),  256), length(grid.active_cells_surface)
+
+@inline active_linear_index_to_interior_tuple(idx, grid::ActiveCellsIBG) = Base.map(Int, grid.active_cells_interior[idx])
+@inline  active_linear_index_to_surface_tuple(idx, grid::ActiveCellsIBG) = Base.map(Int, grid.active_cells_surface[idx])
+
+function ImmersedBoundaryGrid(grid, ib; active_cells_map::Bool = true) 
+
+    ibg = ImmersedBoundaryGrid(grid, ib)
+    TX, TY, TZ = topology(ibg)
+    
     # Create the cells map on the CPU, then switch it to the GPU
     if active_cells_map 
-        map = active_cells_map(grid, ib)
-        map = arch_array(architecture(grid), map)
+        map_interior = active_cells_map_interior(ibg)
+        map_interior = arch_array(architecture(ibg), map_interior)
+        # map_surface = active_cells_map_surface(ibg)
+        # map_surface = arch_array(architecture(ibg), map_surface)
     else
-        map = nothing
+        map_interior = nothing
     end
 
-    return ImmersedBoundaryGrid{TX, TY, TZ}(grid, ib, map)
+    return ImmersedBoundaryGrid{TX, TY, TZ}(ibg.underlying_grid, 
+                                            ibg.immersed_boundary, 
+                                            map_interior)
 end
 
-function compute_active_cells(grid, ib)
-    is_immersed_operation = KernelFunctionOperation{Center, Center, Center}(active_cell, grid)
-    active_cells_field = Field{Center, Center, Center}(grid, Bool)
+@inline active_cell(i, j, k, ibg) = !immersed_cell(i, j, k, ibg)
+@inline active_column(i, j, k, grid, column) = column[i, j, k] != 0
+
+function compute_active_cells_interior(ibg)
+    is_immersed_operation = KernelFunctionOperation{Center, Center, Center}(active_cell, ibg)
+    active_cells_field = Field{Center, Center, Center}(ibg, Bool)
     set!(active_cells_field, is_immersed_operation)
     return active_cells_field
 end
 
+function compute_active_cells_surface(ibg)
+    one_field = ConditionalOperation{Center, Center, Center}(OneField(Int), identity, ibg, NotImmersed(truefunc), 0.0)
```

Comment:
```suggestion
    one_field = ConditionalOperation{Center, Center, Center}(OneField(Int), identity, ibg, NotImmersed(truefunc), 0)
```

to avoid type instability?